tracer/test/Datadog.Trace.Tests/CallTarget/ValueTaskAsyncContinuationGeneratorTests.cs

```diff
@@ -0,0 +1,345 @@
+                };
+            }
+        }
+
+        [Fact]
+        public async Task SuccessGenericKnownTypeTest()
+        {
+            var tcg = new ValueTaskContinuationGenerator<IntegrationWithKnownType, ValueTaskAsyncContinuationGeneratorTests, ValueTask<string>, string>();
+            var state = CallTargetState.GetDefault();
+            var cTask = tcg.SetContinuation(this, GetPreviousTask(), null, in state);
+
+            var rValue = await cTask;
+            Assert.Equal("ReturnValue[Modified]", rValue);
+
+            async ValueTask<string> GetPreviousTask()
+            {
+                await Task.Delay(1000).ConfigureAwait(false);
+                return "ReturnValue";
+            }
+        }
+
+        internal class CustomException : Exception
+        {
+            public CustomException(string message)
+                : base(message)
+            {
+            }
+        }
+
+        internal class CustomCancellationException : OperationCanceledException
+        {
+            public CustomCancellationException(CancellationToken token)
+                : base(token)
+            {
+            }
+        }
+
+        internal class CustomSynchronizationContext : SynchronizationContext
+        {
+            private readonly TaskCompletionSource<bool> _tcs = new();
+
+            public Task Task => _tcs.Task;
+
+            public override void Post(SendOrPostCallback d, object state)
+            {
+                _tcs.TrySetResult(true);
+            }
+
+            public override void Send(SendOrPostCallback d, object state)
+            {
+                _tcs.TrySetResult(true);
+            }
+        }
+
+        internal class PreserveContextTests
+        {
+            [PreserveContext]
+            public static async Task<TReturn> OnAsyncMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, CallTargetState state)
+            {
+                await Task.Delay(1000).ConfigureAwait(false);
+                return returnValue;
+            }
+        }
+
+        internal class IntegrationWithDuckType
+        {
+            public interface IReturnValue
+            {
+                string Value { get; set; }
+            }
+
+            public static async Task<TReturn> OnAsyncMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, CallTargetState state)
+                where TReturn : IReturnValue
+            {
+                await Task.Delay(1000).ConfigureAwait(false);
+                returnValue.Value += "[Modified]";
+                return returnValue;
+            }
+        }
+
+        internal class ReturnValue
+        {
+            public string Value { get; set; }
+        }
+
+        internal class IntegrationWithKnownType
+        {
+            public static async Task<string> OnAsyncMethodEnd<TTarget>(TTarget instance, string returnValue, Exception exception, CallTargetState state)
```

Comment:
I guess this answers my previous question, though I don't follow _how_ it works 😅

Reply:
Something similar to:

```csharp
// The Result of the ContinuationAsync is the Task we return to the method we instrument.
// The difference in this PR is that now we detect `OnAsyncMethodEnd` is return a Task so we do the await.
async Task<string> ContinuationAsync(Task<string> originalReturnTask)
{
     string originalReturnValue = await originalReturnTask;
     string newReturnValue = await OnAsyncMethodEnd(...., originalReturnValue, ....);
     return newReturnValue;
}

async Task<string> Continuation(Task<string> originalReturnTask)
{
     string originalReturnValue = await originalReturnTask;
     string newReturnValue = OnAsyncMethodEnd(...., originalReturnValue, ....);
     return newReturnValue;
}

Task<string> OnMethodEnd(...., Task<string> returnValue, ....)
{
     return ContinuationAsync(returnValue);
}
```